doc/man/gnunet-rsa.1
.TH GNUNET\-RSA 1 "Mar 15, 2012" "GNUnet"

.SH NAME
gnunet\-rsa \- manipulate GNUnet RSA key files

.SH SYNOPSIS
.B gnunet\-rsa
.RI [ options ] FILENAME
.br

.SH DESCRIPTION
\fBgnunet\-rsa\fP can be used to create an RSA private key and to print the corresponding public key.  You must specify a filename containing an RSA private key in GNUnet format as an argument.  If the file does not exist, gnunet\-rsa will create a key.  This may then take a while.  If the option \-p is given, the corresponding public key will be printed to the console.

.SH OPTIONS
.B
.IP "\-g COUNT, \-\-generate-keys=COUNT"
Create COUNT public-private key pairs and write them to FILENAME.  Used for creating a file for testing.
.B
.IP "\-p, \-\-print-public-key"
Print the corresponding public key to stdout.
.B
.IP "\-P, \-\-print-peer-identity"
Print the corresponding peer identity (hash of the public key) to stdout.  This hash is used for the name of peers.
.B
.IP "\-s, \-\-print-short-identity"
Print the corresponding short hash (256-bit hash of the public key) to stdout.  This hash is used for names in the zkey zone.
.B
.IP "\-c FILENAME,  \-\-config=FILENAME"
Use the configuration file FILENAME.
.B
.IP "\-h, \-\-help"
Print short help on options.
.B
.IP "\-L LOGLEVEL, \-\-loglevel=LOGLEVEL"
Use LOGLEVEL for logging.  Valid values are DEBUG, INFO, WARNING and ERROR.
.B
.IP "\-v, \-\-version"
Print GNUnet version number.


.SH BUGS
Report bugs by using Mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>



ChangeLog
Mon Oct 31 18:46:36 CET 2011
	Releasing GNUnet 0.9.0pre4.  This release includes a working
	implementation of mesh (unoptimized), a working implementation
	of the VPN (could use more testing), a working implementation 
	of the WLAN transport (still very slow, unoptimized) and
	many, many bugfixes all over the place.  The major known
	issues are:

	* TESTING is known to sometimes crash on shutdown
	* DV is known to crash / not work properly
	* CHAT is experimental and was not reviewed yet

	More details to individual bugs that were fixed in this
	release are documented at:
	https://gnunet.org/bugs/changelog_page.php?version_id=182

Mon Sep 19 14:16:22 CEST 2011
	Releasing GNUnet 0.9.0pre3.  Instead of listing all that has
	changed, here is the short list of what is known NOT to work,
	in order of minor flaws to major issues, by module:

	* FS still has a few missing features, but largely works
	* TRANSPORT ATS is not used for actual bandwidth assignment,
	          important options to make it truly useful are still missing
	* TESTING does not support incremental HELLO changes, 
	          code has known shutdown issues (can crash)
	* VPN is experimental, but has been shown to work to some degree
	* MESH does not work (mock API emulates for VPN)
        * TRANSPORT-WLAN is experimental
        * TRANSPORT-HTTP is experimental, has known bugs
	* TRANSPORT-DV is experimental, has known bugs, protocol may still change
	* CHAT is experimental, has known bugs and was not reviewed yet

Thu Dec 23 23:33:05 CET 2010
	Releasing GNUnet 0.9.0pre2.

Thu Oct	 7 20:51:05 CEST 2010
	Removed dependency on GNU MP (GMP). The code now uses the libgcrypt MPI
	interface.

Sat Jul  3 20:47:45 CEST 2010
	Releasing GNUnet 0.9.0pre1.

Wed Jun 23 16:34:38 CEST 2010
	Added support for systemd-compatible passing of listen-sockets
	by ARM to services as well as systemd compatibility for gnunet-service-arm
	itself.  At least for non-MINGW systems this should work.

Sat Jun  5 18:08:39 CEST 2010
	Added support for UNIX domain sockets, code also defaults to
	them when available.

Sun May  2 13:49:10 CEST 2010
	Fixed problem with platform-dependence of format for IP addresses
	in HELLOs for TCP and UDP transport.
	Reduced address length field in HELLOs to 16 bit (was 32 bit).
	These changes break transport compatibility.

Fri Apr 16 18:19:05 CEST 2010
	Nearly complete rewrite and new overall architecture.  Many
	features are still missing, but basic system seems to be
	working again.  Maybe time to again track major changes
	in the ChangeLog.  Releasing 0.9.0pre0.


Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent.
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).

Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in
	util/network/dns.c; improved IPv6 support for
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.


src/util/gnunet-rsa.c
/*
     This file is part of GNUnet.
     (C) 2012 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 3, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file util/gnunet-rsa.c
 * @brief tool to manipulate RSA key files
 * @author Christian Grothoff
 */
#include "platform.h"
#include "gnunet_util_lib.h"
#include <gcrypt.h>


/**
 * Flag for printing public key.
 */
static int print_public_key;

/**
 * Flag for printing hash of public key.
 */
static int print_peer_identity;

/**
 * Flag for printing short hash of public key.
 */
static int print_short_identity;

/**
 * Use weak random number generator for key generation.
 */
static int weak_random;

/**
 * Option set to create a bunch of keys at once.
 */
static unsigned int make_keys;

/**
 * The private information of an RSA key pair.
 * NOTE: this must match the definition in crypto_ksk.c and crypto_rsa.c!
 */
struct GNUNET_CRYPTO_RsaPrivateKey
{
  gcry_sexp_t sexp;
};


#if 0
/**
 * Create a new private key. Caller must free return value.
 *
 * @return fresh private key
 */
struct GNUNET_CRYPTO_RsaPrivateKey *
GNUNET_CRYPTO_rsa_key_create ()
{
  struct GNUNET_CRYPTO_RsaPrivateKey *ret;
  gcry_sexp_t s_key;
  gcry_sexp_t s_keyparam;

  GNUNET_assert (0 ==
                 gcry_sexp_build (&s_keyparam, NULL,
                                  "(genkey(rsa(nbits %d)(rsa-use-e 3:257)))",
                                  HOSTKEY_LEN));
  GNUNET_assert (0 == gcry_pk_genkey (&s_key, s_keyparam));
  gcry_sexp_release (s_keyparam);
#if EXTRA_CHECKS
  GNUNET_assert (0 == gcry_pk_testkey (s_key));
#endif
  ret = GNUNET_malloc (sizeof (struct GNUNET_CRYPTO_RsaPrivateKey));
  ret->sexp = s_key;
  return ret;
}
#endif


/**
 * Create a flat file with a large number of key pairs for testing.
 */
static void
create_keys (const char *fn)
{
  time_t start;
  struct GNUNET_HashCode hc;
  struct GNUNET_HashCode h2;
  struct GNUNET_HashCode h3;
  FILE *f;
  struct GNUNET_CRYPTO_RsaPrivateKey *pk;
  struct GNUNET_CRYPTO_RsaPrivateKeyBinaryEncoded *enc;

  start = time (NULL);
  GNUNET_CRYPTO_hash (&start, sizeof (start), &hc);
  if (NULL == (f = fopen (fn, "w+")))
    {
      fprintf (stderr,
	       _("Failed to open `%s': %s\n"),
	       fn,
	       STRERROR (errno));
      return;
    }
  fprintf (stderr,
	   _("Generating %u keys, please wait"),
	   make_keys);
  while (0 < make_keys--)
  {    
    fprintf (stderr,
	     ".");
    GNUNET_CRYPTO_hash (&make_keys, sizeof (make_keys), &h2);
    GNUNET_CRYPTO_hash (&hc, sizeof (hc), &h3);
    GNUNET_CRYPTO_hash_xor (&h2, &h3, &hc);
    if (NULL == (pk = GNUNET_CRYPTO_rsa_key_create_from_hash (&hc)))
    {	
       GNUNET_break (0);
       break;
    }
    enc = GNUNET_CRYPTO_rsa_encode_key (pk);
    if (htons (enc->len) != fwrite (enc, 1, htons (enc->len), f))
      {
	fprintf (stderr,
		 _("\nFailed to write to `%s': %s\n"),
		 fn,
		 STRERROR (errno));
	GNUNET_CRYPTO_rsa_key_free (pk);
	GNUNET_free (enc);
	break;
      }
    GNUNET_CRYPTO_rsa_key_free (pk);
    GNUNET_free (enc);
  }
  if (0 == make_keys)
    fprintf (stderr,
	     _("Finished!\n"));
  fclose (f);
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls, char *const *args, const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  struct GNUNET_CRYPTO_RsaPrivateKey *pk;
  struct GNUNET_CRYPTO_RsaPublicKeyBinaryEncoded pub;
  struct GNUNET_PeerIdentity pid;

  if (NULL == args[0])
  {
    fprintf (stderr, _("No hostkey file specified on command line\n"));
    return;
  }
  if (0 != weak_random)    
    GNUNET_CRYPTO_random_disable_entropy_gathering ();  
  if (make_keys > 0)
  {
    create_keys (args[0]);
    return;
  }
  pk = GNUNET_CRYPTO_rsa_key_create_from_file (args[0]);
  if (NULL == pk)
    return;
  if (print_public_key)
  {
    char *s;

    GNUNET_CRYPTO_rsa_key_get_public (pk, &pub);
    s = GNUNET_CRYPTO_rsa_public_key_to_string (&pub);
    fprintf (stdout, "%s\n", s);
    GNUNET_free (s);
  }
  if (print_peer_identity)
  {
    struct GNUNET_CRYPTO_HashAsciiEncoded enc;

    GNUNET_CRYPTO_rsa_key_get_public (pk, &pub);
    GNUNET_CRYPTO_hash (&pub, sizeof (pub), &pid.hashPubKey);
    GNUNET_CRYPTO_hash_to_enc (&pid.hashPubKey, &enc);
    fprintf (stdout, "%s\n", enc.encoding);
  }
  if (print_short_identity)
  {
    struct GNUNET_CRYPTO_ShortHashAsciiEncoded enc;
    struct GNUNET_CRYPTO_ShortHashCode sh;

    GNUNET_CRYPTO_rsa_key_get_public (pk, &pub);
    GNUNET_CRYPTO_short_hash (&pub, sizeof (pub), &sh);
    GNUNET_CRYPTO_short_hash_to_enc (&sh, &enc);
    fprintf (stdout, "%s\n", enc.short_encoding);
  }
  GNUNET_CRYPTO_rsa_key_free (pk);
}


/**
 * Program to manipulate RSA key files.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  static const struct GNUNET_GETOPT_CommandLineOption options[] = {
    { 'g', "generate-keys", "COUNT",
      gettext_noop ("create COUNT public-private key pairs (for testing)"),
      1, &GNUNET_GETOPT_set_uint, &make_keys },
    { 'p', "print-public-key", NULL,
      gettext_noop ("print the public key in ASCII format"),
      0, &GNUNET_GETOPT_set_one, &print_public_key },
    { 'P', "print-peer-identity", NULL,
      gettext_noop ("print the hash of the public key in ASCII format"),
      0, &GNUNET_GETOPT_set_one, &print_peer_identity },
    { 's', "print-short-identity", NULL,
      gettext_noop ("print the short hash of the public key in ASCII format"),
      0, &GNUNET_GETOPT_set_one, &print_short_identity },
    { 'w', "weak-random", NULL,
      gettext_noop ("use insecure, weak random number generator for key generation (for testing only)"),
      0, &GNUNET_GETOPT_set_one, &weak_random },
    GNUNET_GETOPT_OPTION_END
  };

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  return (GNUNET_OK ==
          GNUNET_PROGRAM_run (argc, argv, "gnunet-rsa [OPTIONS] keyfile",
                              gettext_noop ("Manipulate GNUnet private RSA key files"),
                              options, &run, NULL)) ? 0 : 1;
}

/* end of gnunet-rsa.c */
